sound/soc/sof/virtio-be.c

```diff
@@ -79,10 +85,328 @@ static int sbe_ipc_comp(struct snd_sof_dev *sdev, int vm_id,
 	return 0;
 }
 
-/* TODO: to be implement later */
+/*
+ * This function is to get the BE dai link for the GOS
+ * It uses the dai_link name to find the BE dai link.
+ * The Current dai_link name "vm_dai_link" is for the GOS,
+ * which means only one Virtual Machine is supported.
+ * And the VM only support one playback pcm and one capture pcm.
+ * After we switch to the new topology, we can support multiple
+ * VMs and multiple PCM streams for each VM.
+ * This function may be abandoned after switching to the new topology.
+ */
+static struct snd_pcm_substream *
+sbe_get_substream(struct snd_sof_dev *sdev,
+		  struct snd_soc_pcm_runtime *rtd, int direction)
+{
+	struct snd_pcm *pcm;
+	struct snd_pcm_str *stream;
+	struct snd_pcm_substream *substream = NULL;
+	struct snd_soc_dai_link *dai_link;
+	struct snd_soc_card *card = sdev->card;
+
+	list_for_each_entry(rtd, &card->rtd_list, list) {
+		pcm = rtd->pcm;
+		if (!pcm)
+			continue;
+
+		stream = &pcm->streams[direction];
+		if (!stream)
+			continue;
+
+		substream = stream->substream;
+		if (substream) {
+			dai_link = rtd->dai_link;
+			if (strcmp(dai_link->name, "vm_dai_link") == 0)
+				return substream;
+		}
+	}
+
+	return NULL;
+}
+
+static int sbe_pcm_open(struct snd_sof_dev *sdev,
+			void *ipc_data, int vm_id)
+{
+	/*
```

Comment:
Stick this comment before func and comment each non obvious block in every function. i.e. I need to know what each block does.